### PR TITLE
RSE-1412: Duplicate link functionality

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
As part of this PR, the Duplicate Link functionality has been implemented in Workflow list page.

## How it looks?
after

## Technical Details
1. Added `WorkflowDuplicateController`, which handles the Duplicate Wokflow feature.
2. The above controller broadcasts `workflow::list::refresh` event, which is caught by `workflowListController` and it refreshes the list.
3. There was a bug in `civicaseCrmApi`, it was not rejecting the promise returned by `crmApi`. This has been fixed. Also unit tests has been added for `civicaseCrmApi`, and many other files spec using `civicaseCrmApi` has been updated.

### CODE REVIEW

File: ang/test/civicase-base/services/crm-api.service.spec.js [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+        it('calls the backend with for all the requested information', () => {
+          expect(crmApiMock.calls.mostRecent().args[0]).toEqual([
+            ['SomeEntity', 'someendpoint', {}],
+            ['SomeEntity2', 'someendpoint2', {}]
+          ]);
+        });
+
+        it('throws an exception', () => {
+          expect(isErrorThrown).toBe(true);
```
Comment (reneolivo) [inline ang/test/civicase-base/services/crm-api.service.spec.js:46]:
```js
errorHandler = jasmine.createSpy('errorHandler');

// ...
.catch(errorHandler);

// ...

it('throws an exception', () => {
  expect(errorHandler).toHaveBeenCalledWith(expectedException);
});
```

File: ang/test/civicase-base/services/crm-api.service.spec.js [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+        it('calls the backend with for all the requested information', () => {
+          expect(crmApiMock.calls.mostRecent().args[0]).toEqual([
+            ['SomeEntity', 'someendpoint', {}],
+            ['SomeEntity2', 'someendpoint2', {}]
+          ]);
+        });
+
+        it('does not throw an exception', () => {
+          expect(isErrorThrown).toBe(false);
```
Comment (reneolivo) [inline ang/test/civicase-base/services/crm-api.service.spec.js:76]:
```js
expect(errorHandler).not.toHaveBeenCalled();
```

File: ang/test/workflow/directives/workflow-list.directive.spec.js [hunk 3]
```diff
@@ -46,6 +46,27 @@
...
+        }));
+        initController();
+        $scope.$digest();
+        $scope.workflows = [];
+        $rootScope.$broadcast('workflow::list::refresh');
+        $scope.$digest();
+      });
+
+      it('fetches the case types for the current case type category', () => {
+        expect(civicaseCrmApiMock.calls.count()).toBe(2);
```
Comment (reneolivo) [inline ang/test/workflow/directives/workflow-list.directive.spec.js:62]: we need to test for the specific calls.

File: ang/workflow/action-links/controllers/workflow-duplicate.controller.js [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+      });
+    }
+
+    /**
+     * Generate workflow `name` from the sent text
+     *
+     * @param {string} text string from which name would be generated
+     * @returns {string} name of the workflow
+     */
+    function generateWorkflowName (text) {
```
Comment (reneolivo) [inline ang/workflow/action-links/controllers/workflow-duplicate.controller.js:127]: shouldn't this be generated by the API?

**Review by reneolivo — CHANGES_REQUESTED** [review]

File: ang/workflow/action-links/controllers/workflow-duplicate.controller.js (hunk 4 above)
Comment (deb1990) [reply ang/workflow/action-links/controllers/workflow-duplicate.controller.js:127]: The original logic was added on the FE by core, and not in api level. So following the same structure
https://github.com/civicrm/civicrm-core/blob/master/ang/crmCaseType.js#L701

Comment (reneolivo) [reply ang/workflow/action-links/controllers/workflow-duplicate.controller.js:127]: *sad noises*

**Review by reneolivo — APPROVED** [review]

**Pull request merged**